### ISSUE TITLE
Errors trying to build TRT Onnx related samples

### ISSUE THREAD

**Issue by bpinaya** [onnx/onnx-tensorrt#75]
Hi there, on TensorRT 5.0.2GA and onnx-tensorrt installed from source I've encountered the following errors while trying to build samples related to onnx. To be more specific, while running `make` on sampleINT8API I get:
```
../Makefile.config:5: CUDA_INSTALL_DIR variable is not specified, using /usr/local/cuda by default, use CUDA_INSTALL_DIR=<cuda_directory> to change.
../Makefile.config:8: CUDNN_INSTALL_DIR variable is not specified, using $CUDA_INSTALL_DIR by default, use CUDNN_INSTALL_DIR=<cudnn_directory> to change.
:
Compiling: sampleINT8API.cpp
sampleINT8API.cpp: In member function ‘bool sampleINT8API::build()’:
sampleINT8API.cpp:448:102: error: cannot convert ‘nvinfer1::INetworkDefinition’ to ‘nvinfer1::INetworkDefinition*’ for argument ‘1’ to ‘nvonnxparser::IParser* nvonnxparser::{anonymous}::createParser(nvinfer1::INetworkDefinition*, nvinfer1::ILogger&)’
     auto parser = SampleUniquePtr<nvonnxparser::IParser>(nvonnxparser::createParser(*network, gLogger));
                                                                                                      ^
../Makefile.config:172: recipe for target '../../bin/dchobj/sampleINT8API.o' failed
make: *** [../../bin/dchobj/sampleINT8API.o] Error 1
```
And while trying to build sampleOnnxMNIST I get:
```
../Makefile.config:5: CUDA_INSTALL_DIR variable is not specified, using /usr/local/cuda by default, use CUDA_INSTALL_DIR=<cuda_directory> to change.
../Makefile.config:8: CUDNN_INSTALL_DIR variable is not specified, using $CUDA_INSTALL_DIR by default, use CUDNN_INSTALL_DIR=<cudnn_directory> to change.
:
Compiling: sampleOnnxMNIST.cpp
sampleOnnxMNIST.cpp: In function ‘void onnxToTRTModel(const string&, unsigned int, nvinfer1::IHostMemory*&)’:
sampleOnnxMNIST.cpp:45:63: error: cannot convert ‘nvinfer1::INetworkDefinition’ to ‘nvinfer1::INetworkDefinition*’ for argument ‘1’ to ‘nvonnxparser::IParser* nvonnxparser::{anonymous}::createParser(nvinfer1::INetworkDefinition*, nvinfer1::ILogger&)’
     auto parser = nvonnxparser::createParser(*network, gLogger);
                                                               ^
../Makefile.config:172: recipe for target '../../bin/dchobj/sampleOnnxMNIST.o' failed
make: *** [../../bin/dchobj/sampleOnnxMNIST.o] Error 1

```
I've downloaded the models and copied them to the required folder but the error persists. 
Anything I might be missing out?
I'm on Ubuntu16.04 with CUDA10 and cudnn 7.4.1.5-1+cuda10.0.

**Comment by bpinaya** [comment]
After a clean install (TRT5GA + CUDA10) the samples seem to be built correctly, after cloning this repo (to implement some ports myself) and running:
```
mkdir build
cd build
cmake .. -DTENSORRT_ROOT=/opt/tensorrt
make -j8
sudo make install
```
the onnx related samples are no longer building, with the log below
```
g++ -MM -MF ../../bin/dchobj/sampleINT8API.d -MP -MT ../../bin/dchobj/sampleINT8API.o -Wall -std=c++11 -I"/usr/local/cuda/include" -I"/usr/local/include" -I"../include" -I"../common" -I"/usr/local/cuda/include" -I"../../include"  -D_REENTRANT sampleINT8API.cpp
Compiling: sampleINT8API.cpp
g++ -Wall -std=c++11 -I"/usr/local/cuda/include" -I"/usr/local/include" -I"../include" -I"../common" -I"/usr/local/cuda/include" -I"../../include"  -D_REENTRANT -g -c -o ../../bin/dchobj/sampleINT8API.o sampleINT8API.cpp
sampleINT8API.cpp: In member function 'bool sampleINT8API::build()':
sampleINT8API.cpp:448:102: error: cannot convert 'nvinfer1::INetworkDefinition' to 'nvinfer1::INetworkDefinition*' for argument '1' to 'nvonnxparser::IParser* nvonnxparser::{anonymous}::createParser(nvinfer1::INetworkDefinition*, nvinfer1::ILogger&)'
     auto parser = SampleUniquePtr<nvonnxparser::IParser>(nvonnxparser::createParser(*network, gLogger));
                                                                                                      ^
../Makefile.config:172: recipe for target '../../bin/dchobj/sampleINT8API.o' failed
make[1]: *** [../../bin/dchobj/sampleINT8API.o] Error 1
make[1]: Leaving directory '/workspace/tensorrt/samples/sampleINT8API'
Makefile:37: recipe for target 'all' failed
make: *** [all] Error 2
```
When I pass `-DTENSORRT_ROOT=/opt/tensorrt` to the cmake it's where trt was installed, a `ls` of that dir results in:
```
bin  data  python  samples
```
Any idea what am I doing wrong?

**Comment by yinghai** [comment]
@benbarsdell This is a bit weird because I don't remember that we have such a sample in our repo.

**Comment by bpinaya** [comment]
@yinghai I'm referring to the samples provided by TensorRT regarding Onnx [here](https://docs.nvidia.com/deeplearning/sdk/tensorrt-developer-guide/index.html#onnx_mnist_sample).
I want to implement support for a couple onnx layers but I'm having some trouble getting it to work together.

**Comment by bpinaya** [comment]
Ok I tried with TensorRT4 and it works correctly, even TensorRT5RC, this issue is present on TensorRT5.0.2GA.

**Comment by bpinaya** [comment]
@yinghai I think I need to explain myself better, so this is my problem:
I have a trained network already in onnx. I want to port that network to TRT, with the default instalation of TensorRT5.0.2GA I can load the network using the cpp api, but I'm lacking support for some layers and that's where the problem emerges. According to my understanding, to add support for onnx layers we need to clone this repo, build and install it right? And that should replace the onnx parser shipped with TensorRT? When building I have no problems, but when trying to rebuild the samples shipped with TRT I get this:
```
../Makefile.config:5: CUDA_INSTALL_DIR variable is not specified, using /usr/local/cuda by default, use CUDA_INSTALL_DIR=<cuda_directory> to change.
../Makefile.config:8: CUDNN_INSTALL_DIR variable is not specified, using $CUDA_INSTALL_DIR by default, use CUDNN_INSTALL_DIR=<cudnn_directory> to change.
:
Compiling: sampleINT8API.cpp
sampleINT8API.cpp: In member function ‘bool sampleINT8API::build()’:
sampleINT8API.cpp:448:102: error: cannot convert ‘nvinfer1::INetworkDefinition’ to ‘nvinfer1::INetworkDefinition*’ for argument ‘1’ to ‘nvonnxparser::IParser* nvonnxparser::{anonymous}::createParser(nvinfer1::INetworkDefinition*, nvinfer1::ILogger&)’
     auto parser = SampleUniquePtr<nvonnxparser::IParser>(nvonnxparser::createParser(*network, gLogger));
                                                                                                      ^
../Makefile.config:172: recipe for target '../../bin/dchobj/sampleINT8API.o' failed
make: *** [../../bin/dchobj/sampleINT8API.o] Error 1
```
If I'm confusing things I'm sorry, but according to my understanding in order to add support for other layers we need to do so with this repo, I'm attaching this screenshot (taken from the dev-guide) from where I got to that conclusion.
![image](https://user-images.githubusercontent.com/2473733/49009846-dca1f480-f171-11e8-8b72-bc6f945ca961.png)

**Comment by yinghai** [comment]
@bpinaya IIRC, onnx import code in this repo is not exactly the same as that in TensorRT 5.0. I have no access to code in TensorRT 5.0 so I'm not sure. For your case, if you want to implement some layers within this repo and still use TensorRT, I think it's still doable. You'll need the onnx2trt binary to do your conversion test (https://github.com/onnx/onnx-tensorrt/blob/master/CMakeLists.txt#L228). It will read in an onnx file, do the conversion and write out a trt engine file. With that serialized engine file, you can just use TensorRT to deserialize it and run it (https://docs.nvidia.com/deeplearning/sdk/tensorrt-developer-guide/index.html#serial_model_c).

**Comment by benbarsdell** [comment]
@yinghai it appears the difference in API arose from here:
https://github.com/onnx/onnx-tensorrt/blame/master/NvOnnxParser.h#L197
https://github.com/onnx/onnx-tensorrt/commit/fa0964e8477fc004ee2f49ee77ffce0bf7f711a9#diff-5ac4e27fe04b3593a9d33ef2beeec61fR197
And this mustn't have been sync'd properly into the TRT5 package.
Before I follow up, could you clarify why the breaking change was necessary? I must not have noticed it at the time.

**Comment by yinghai** [comment]
@benbarsdell I remember what happened. We used to pass `nvinfer1::INetworkDefinition*` as pointer when `createParser`. Then it got changed to passing by reference and I changed it back. The reason is that during ONNXIFI interface, we wanted to create a parser without `nvinfer1::INetworkDefinition`, just for some query purpose (e.g. `onnxGetCompatibility`). Therefore passing `nullptr` is more convenient. Does that make sense?
